api/app.go

```diff
@@ -0,0 +1,295 @@
+
+	responseData["redirect"] = callback + "?code=" + url.QueryEscape(authData.Code) + "&state=" + url.QueryEscape(authData.State)
+
+	w.Write([]byte(model.MapToJson(responseData)))
+}
+
+func getAccessToken(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.LogAudit("attempt")
+
+	props := model.MapFromJson(r.Body)
+
+	grantType := props["grant_type"]
+	if grantType != model.ACCESS_TOKEN_GRANT_TYPE {
+		c.Err = model.NewOAuthAppError("getAccessToken", "invalid_request", "Bad grant_type")
+		return
+	}
+
+	clientId := props["client_id"]
+	if len(clientId) != 26 {
+		c.Err = model.NewOAuthAppError("getAccessToken", "invalid_request", "Bad client_id")
+		return
+	}
+
+	secret := props["client_secret"]
+	if len(secret) == 0 {
+		c.Err = model.NewOAuthAppError("getAccessToken", "invalid_request", "Missing client_secret")
+		return
+	}
+
+	code := props["code"]
+	if len(code) == 0 {
+		c.Err = model.NewOAuthAppError("getAccessToken", "invalid_request", "Missing code")
+		return
+	}
+
+	redirectUri := props["redirect_uri"]
+
+	achan := Srv.Store.App().Get(clientId)
+	tchan := Srv.Store.AccessData().GetByAuthCode(code)
+
+	var authData *model.AuthData
+	if ad, ok := authCodeCache.Get(code); ok {
+		authData = ad.(*model.AuthData)
+	}
+
+	if authData == nil {
+		if result := <-Srv.Store.AuthData().Get(code); result.Err != nil {
+			c.Err = model.NewOAuthAppError("getAccessToken", "invalid_grant", "Invalid or expired authorization code")
+			return
+		} else {
+			authData = result.Data.(*model.AuthData)
+		}
+	}
+
+	if authData == nil || authData.IsExpired() {
+		c.Err = model.NewOAuthAppError("getAccessToken", "invalid_grant", "Invalid or expired authorization code")
+		return
+	}
+
+	if authData.RedirectUri != redirectUri {
+		c.Err = model.NewOAuthAppError("getAccessToken", "invalid_request", "Supplied redirect_uri does not match authorization code redirect_uri")
+		return
+	}
+
+	if !model.ComparePassword(code, fmt.Sprintf("%v:%v:%v:%v", clientId, redirectUri, authData.CreateAt, authData.UserId)) {
+		c.Err = model.NewOAuthAppError("getAccessToken", "invalid_grant", "Invalid or expired authorization code")
+		return
+	}
+
+	var app *model.App
+	if result := <-achan; result.Err != nil {
+		c.Err = model.NewOAuthAppError("getAccessToken", "invalid_client", "Invalid client credentials")
+		return
+	} else {
+		app = result.Data.(*model.App)
+	}
+
+	if !model.ComparePassword(app.ClientSecret, secret) {
+		c.Err = model.NewOAuthAppError("getAccessToken", "invalid_client", "Invalid client credentials")
+		return
+	}
+
+	callback := redirectUri
+	if len(callback) == 0 {
+		callback = app.CallbackUrl
+	}
+
+	if result := <-tchan; result.Err != nil {
+		c.Err = model.NewOAuthAppError("getAccessToken", "server_error", "Encountered internal server error while accessing database")
+		return
+	} else if result.Data != nil {
+		accessData := result.Data.(*model.AccessData)
+		token, _ := model.AesDecrypt(utils.Cfg.ServiceSettings.AesKey, accessData.Token)
```

Comment:
Yes, I actually forgot we had the sql_store.go decrypt and encrypt functions. Might be able to just get rid of one set of the functions